### DISCUSSION TITLE
import/read events from a CSV file

### DISCUSSION THREAD

**Discussion by Argubi** [kineticstoolkit/kineticstoolkit#221] [1. Users Q&A]
Hi, I recently discoverd ktk and it looks really promising. I followed your instructional but could not for the life of me discover how to read events, that are already part of a .csv file that I import as a dataframa and afterwards transfer into a Timeseries Object.
I have EMgData which is exported as a CSV file and Gait events (heelstrike/toe off) are stored in one (or two) columns in the csv export as "On" and "Off" (strings). I would like to import these as events. As of now they are part of the array which contains the emg data.
Am I missing something?
All the best and thank you for your response!

**Comment by felixchenier** [comment]
Hi @Argubi 
I'm sure we'll make it work. Each csv file is different, so there could not be a standard way to read it. I think you are already well started by reading the data in a DataFrame. For the events, the trick would be to parse the events using a for loop that calls TimeSeries.add_event. Maybe you could post the first few lines of your CSV file so that I can understand how it is structured, and the code you tried so far? We can then start from this.

**Reply by Argubi** [reply]
[Example.csv](https://github.com/kineticstoolkit/kineticstoolkit/files/14066789/Example.csv)
Hi, thank you for the super fast response. The CSV file looks like this...
Greetings

**Reply by Argubi** [reply]
My very first approach as workaround to see if I can create events according to the file (without a loop) looked like this:

```
import kineticstoolkit.lab as ktk
import pandas as pd

df =    pd.read_csv(("Example.csv"),index_col="Time (s):")

on_indices = df[df.iloc[:, -1] == "On"].index
off_indices = df[df.iloc[:, -1] == "Off"].index

ts = ktk.TimeSeries.from_dataframe(df)

ts = ts.add_event(on_indices[0], "On")
ts = ts.add_event(off_indices[0], "Off")

**Reply by felixchenier** [reply]
@Argubi I think you managed very well. Your code works almost perfectly on my side, I had to make two small modifications:

1. Specify that the separator is `;` in read_csv.
2. Convert the data to numeric in the DataFrame. For some reasons, some of your data are not numbers, and therefore these columns are read as strings instead of numbers. I think your process for generating the CSV has a bug, since we see dates (?) instead of data at some places in the file. This may explain the numerous erroneous data in the plot.

On my side, running this code

```{python}
import kineticstoolkit.lab as ktk
import pandas as pd

df = pd.read_csv(("Example.csv"), index_col="Time (s):", sep=";")

on_indices = df[df.iloc[:, -1] == "On"].index
off_indices = df[df.iloc[:, -1] == "Off"].index

for column in df.columns:
    df[column] = pd.to_numeric(df[column], 'coerce')

ts = ktk.TimeSeries.from_dataframe(df)

ts = ts.add_event(on_indices[0], "On")
ts = ts.add_event(off_indices[0], "Off")

ts.plot()
```

gives this figure:

<img width="641" alt="image" src="https://github.com/kineticstoolkit/kineticstoolkit/assets/34967663/1fe2df6f-845c-42d2-a27f-b22c308815cc">
